gui/workorder-finetune

```diff
@@ -0,0 +1,544 @@
+            row_height = 4,
+        },
+        widgets.Label{
+            frame = { l = 0, b = 0 },
+            text = {
+                { key = 'LEAVESCREEN', text = ': Back',
+                  on_activate = self:callback('dismiss') }
+            }
+        },
+    }
+
+    self:initListChoices()
+end
+
+function JobDetails:onGetSelectedJob()
+    return self.job
+end
+
+describe_item_type = wsj.describe_item_type
+is_caste_mat = wsj.is_caste_mat
+describe_material = wsj.describe_material
+list_flags = wsj.list_flags
+
+function describe_item_traits(iobj)
+    local line1 = {}
+    local reaction = df.reaction.find(iobj.reaction_id)
+    if reaction and #iobj.contains > 0 then
+        for _,ri in ipairs(iobj.contains) do
+            table.insert(line1, 'has '..utils.call_with_string(
+                reaction.reagents[ri],'getDescription',iobj.reaction_id
+            ))
+        end
+    end
+    if iobj.metal_ore >= 0 then
+        local ore = dfhack.matinfo.decode(0, iobj.metal_ore)
+        if ore then
+            table.insert(line1, 'ore of '..ore:toString())
+        end
+    end
+    if iobj.has_material_reaction_product ~= '' then
+        table.insert(line1, iobj.has_material_reaction_product .. '-producing')
+    end
+    if iobj.reaction_class ~= '' then
+        table.insert(line1, 'reaction class '..iobj.reaction_class)
+    end
+    if iobj.has_tool_use >= 0 then
+        table.insert(line1, 'has use '..df.tool_uses[iobj.has_tool_use])
+    end
+
+    list_flags(line1, iobj.flags1)
+    list_flags(line1, iobj.flags2)
+    list_flags(line1, iobj.flags3)
+
+    if #line1 == 0 then
+        table.insert(line1, 'no traits')
+    end
+    return table.concat(line1, ', ')
+end
+
+function JobDetails:initListChoices()
+    if not self.job.items then
+        self.subviews.list:setChoices({})
+        return
+    end
+
+    local choices = {}
+    for i,iobj in ipairs(self.job.items) do
+        local head = 'Item '..(i+1)..' x'..iobj.quantity
+        if iobj.min_dimension > 0 then
+            head = head .. '(size '..iobj.min_dimension..')'
+        end
+
+        table.insert(choices, {
+            index = i,
+            iobj = iobj,
+            text = {
+                head, NEWLINE,
+                '  ', { text = curry(describe_item_type, iobj) }, NEWLINE,
+                '  ', { text = curry(describe_material, iobj) }, NEWLINE,
+                '  ', { text = curry(describe_item_traits, iobj) }, NEWLINE
+            }
+        })
+    end
+
+    self.subviews.list:setChoices(choices)
+end
+
+JobDetails.canChangeIType = wsj.JobDetails.canChangeIType
+JobDetails.setItemType = wsj.JobDetails.setItemType
+JobDetails.onChangeIType = wsj.JobDetails.onChangeIType
+JobDetails.canChangeMat = wsj.JobDetails.canChangeMat
+JobDetails.setMaterial = wsj.JobDetails.setMaterial
+JobDetails.findUnambiguousItem = wsj.JobDetails.findUnambiguousItem
+JobDetails.onChangeMat = wsj.JobDetails.onChangeMat
```

Comment:
Stealing individual methods from a class, while technically functional, could lead to issues later if `wsj` is updated with undesired behavior. Could you refactor `wsj.JobDetails` so that the common functions are in a superclass? Then that class could be clearly marked as "shared" and both the `JobDetails` here and the `JobDetails` in `wsj` could inherit from that common class.

Reply:
Is there a convention how such a shared class should look like? I think to call it `JobDetailsBase` and place in `/scripts/gui/`.

Reply:
`JobDetailsBase` sounds good. It can either stay in wsj (commented to indicate that it is shared across scripts) or it can go in `internal/gui/`. It shouldn't go in a new file in `gui/` because then it would show up in `ls` as an independent script

Reply:
so, I started to actually do this and it would be a **huge** refactoring. The problem is one class inherits from `MenuOverlay` while another from `FramedScreen`. We need multi-inheritance or to change both scripts in such a way most of the rendering happens inside a widget, then `JobDetailsBase` can inherit from `Widget`. But this is IMO too much work for little benefit. Should we actually have the class inheritance in place, any change in the base class that's undesired in one of the scripts will still need to be addressed. The only thing we win is obviously marking those functions as `shared` - maybe we could just leave a comment? "Beware: this script shares functions with `worder-details.lua`" or something.